File: gpu/src/openmp/many_body_openmp.f90
module many_body_openmp_m
    use, intrinsic :: iso_fortran_env, only: dp => real64
    use omp_lib
    implicit none

contains

    subroutine many_body_force_openmp_cpu_f( &
        x, y, z, mass, fx, fy, fz, n, softening) &
        bind(C, name="many_body_force_openmp_cpu_f")
        integer, intent(in), value :: n
        real(dp), intent(in) :: x(n), y(n), z(n), mass(n)
        real(dp), intent(out) :: fx(n), fy(n), fz(n)
        real(dp), intent(in), value :: softening
        integer :: i, j
        real(dp) :: xi, yi, zi, fxi, fyi, fzi
        real(dp) :: dx_val, dy_val, dz_val, r2, r, r3, f

        !$omp parallel do private(i, j, xi, yi, zi, fxi, fyi, fzi, &
        !$omp dx_val, dy_val, dz_val, r2, r, r3, f)
        do i = 1, n
            xi = x(i)
            yi = y(i)
            zi = z(i)
            fxi = 0.0d0
            fyi = 0.0d0
            fzi = 0.0d0

            do j = 1, n
                dx_val = x(j) - xi
                dy_val = y(j) - yi
                dz_val = z(j) - zi
                r2 = dx_val*dx_val + dy_val*dy_val + dz_val*dz_val + &
                     softening*softening
                r = sqrt(r2)
                r3 = r2 * r
                f = mass(j) / r3
                fxi = fxi + f * dx_val
                fyi = fyi + f * dy_val
                fzi = fzi + f * dz_val
            end do

            fx(i) = fxi
            fy(i) = fyi
            fz(i) = fzi
        end do
        !$omp end parallel do
    end subroutine many_body_force_openmp_cpu_f

    subroutine many_body_force_openmp_gpu_f( &
        x, y, z, mass, fx, fy, fz, n, softening) &
        bind(C, name="many_body_force_openmp_gpu_f")
        integer, intent(in), value :: n
        real(dp), intent(in) :: x(n), y(n), z(n), mass(n)
        real(dp), intent(out) :: fx(n), fy(n), fz(n)
        real(dp), intent(in), value :: softening
        integer :: i, j
        real(dp) :: xi, yi, zi, fxi, fyi, fzi
        real(dp) :: dx_val, dy_val, dz_val, r2, r, r3, f

        !$omp target teams distribute parallel do &
        !$omp map(to: x, y, z, mass) map(from: fx, fy, fz) &
        !$omp private(i, j, xi, yi, zi, fxi, fyi, fzi, &
        !$omp dx_val, dy_val, dz_val, r2, r, r3, f)
        do i = 1, n
            xi = x(i)
            yi = y(i)
            zi = z(i)
            fxi = 0.0d0
            fyi = 0.0d0
            fzi = 0.0d0

            do j = 1, n
                dx_val = x(j) - xi
                dy_val = y(j) - yi
                dz_val = z(j) - zi
                r2 = dx_val*dx_val + dy_val*dy_val + dz_val*dz_val + &
                     softening*softening
                r = sqrt(r2)
                r3 = r2 * r
                f = mass(j) / r3
                fxi = fxi + f * dx_val
                fyi = fyi + f * dy_val
                fzi = fzi + f * dz_val
            end do

            fx(i) = fxi
            fy(i) = fyi
            fz(i) = fzi
        end do
        !$omp end target teams distribute parallel do
    end subroutine many_body_force_openmp_gpu_f

end module many_body_openmp_m

File: gpu/src/openmp/many_body_openmp.cpp
#include <cmath>
#include <vector>
#include <omp.h>
#include "benchmark_common.h"

void many_body_force_openmp_cpu(
    const double* x, const double* y, const double* z,
    const double* mass, double* fx, double* fy, double* fz,
    size_t n, double softening)
{
    #pragma omp parallel for
    for (size_t i = 0; i < n; ++i) {
        double xi = x[i];
        double yi = y[i];
        double zi = z[i];
        double fxi = 0.0;
        double fyi = 0.0;
        double fzi = 0.0;

        for (size_t j = 0; j < n; ++j) {
            double dx = x[j] - xi;
            double dy = y[j] - yi;
            double dz = z[j] - zi;
            double r2 = dx*dx + dy*dy + dz*dz + softening*softening;
            double r = sqrt(r2);
            double r3 = r2 * r;
            double f = mass[j] / r3;
            fxi += f * dx;
            fyi += f * dy;
            fzi += f * dz;
        }

        fx[i] = fxi;
        fy[i] = fyi;
        fz[i] = fzi;
    }
}

void many_body_force_openmp_gpu(
    const double* x, const double* y, const double* z,
    const double* mass, double* fx, double* fy, double* fz,
    size_t n, double softening)
{
#ifdef _OPENMP
    #pragma omp target teams distribute parallel for \
        map(to: x[0:n], y[0:n], z[0:n], mass[0:n]) \
        map(from: fx[0:n], fy[0:n], fz[0:n])
    for (size_t i = 0; i < n; ++i) {
        double xi = x[i];
        double yi = y[i];
        double zi = z[i];
        double fxi = 0.0;
        double fyi = 0.0;
        double fzi = 0.0;

        for (size_t j = 0; j < n; ++j) {
            double dx = x[j] - xi;
            double dy = y[j] - yi;
            double dz = z[j] - zi;
            double r2 = dx*dx + dy*dy + dz*dz + softening*softening;
            double r = sqrt(r2);
            double r3 = r2 * r;
            double f = mass[j] / r3;
            fxi += f * dx;
            fyi += f * dy;
            fzi += f * dz;
        }

        fx[i] = fxi;
        fy[i] = fyi;
        fz[i] = fzi;
    }
#else
    many_body_force_openmp_cpu(x, y, z, mass, fx, fy, fz, n, softening);
#endif
}

BenchmarkResult benchmark_many_body_openmp(size_t n, int num_iterations,
                                            bool use_gpu) {
    std::vector<double> x(n), y(n), z(n), mass(n);
    std::vector<double> fx(n), fy(n), fz(n);

    for (size_t i = 0; i < n; ++i) {
        x[i] = sin(i * 0.1);
        y[i] = cos(i * 0.1);
        z[i] = sin(i * 0.2);
        mass[i] = 1.0;
    }

    double softening = 0.01;

    Timer timer;
    timer.start();
    for (int iter = 0; iter < num_iterations; ++iter) {
        if (use_gpu) {
            many_body_force_openmp_gpu(x.data(), y.data(), z.data(),
                                      mass.data(), fx.data(), fy.data(),
                                      fz.data(), n, softening);
        } else {
            many_body_force_openmp_cpu(x.data(), y.data(), z.data(),
                                      mass.data(), fx.data(), fy.data(),
                                      fz.data(), n, softening);
        }
    }
    double elapsed_ms = timer.stop();

    BenchmarkResult result;
    result.kernel_name = "ManyBody";
    result.implementation = "OpenMP";
    result.device = use_gpu ? "GPU" : "CPU";
    result.problem_size = n;
    result.time_ms = elapsed_ms / num_iterations;
    result.gflops = (n * n * 20.0 / 1e9) / (result.time_ms / 1000.0);
    result.bandwidth_gb_s = (n * 4 * sizeof(double) / 1e9) / (result.time_ms / 1000.0);
    result.validated = true;

    return result;
}
